### PR TITLE
fix(cdk): fix a bug with CodePipelineSource to create nested repositories

### LINKED ISSUE #27504: pipelines: unable to reference GitLab repositories nested under a group with CodePipelieneSource.connection
### Describe the bug

GitLab repositories that are nested under at least one group cannot be directly with `CodePipelineSource.connection` as it limits repositories to the form `<owner>/<repo>`. GitLab includes the group names in the repository string, falling outside this format (i.e. `owner/group1/group2/groupN/repo`).

### Expected Behavior

I expect the pipeline to synthesize successfully when sourcing from a `CodePipelineSource.connection` that references a GitLab repository nested under multiple groups.

### Current Behavior

The following error is thrown:
```
Error: CodeStar repository name should be a resolved string like '<owner>/<repo>', got 'owner/group1/group2/repo'
```

This error is from [`aws-cdk-lib/pipelines/lib/codepipeline/codepipeline-source.ts` line 429](https://github.com/aws/aws-cdk/blob/b2a895ef285e5451e64c21e179172e998c479582/packages/aws-cdk-lib/pipelines/lib/codepipeline/codepipeline-source.ts#L429).

### Reproduction Steps

```typescript
import * as cdk from 'aws-cdk-lib';
import { CodePipeline, CodePipelineSource, ShellStep } from 'aws-cdk-lib/pipelines';
import { Construct } from 'constructs';

class DummyStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);
  }
}

class DummyApplication extends cdk.Stage {
  constructor(scope: Construct, id: string, props?: cdk.StageProps) {
    super(scope, id, props);

    new DummyStack(this, 'Dummy');
  }
}

export class PipelineStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const pipeline = new CodePipeline(this, 'Pipeline', {
      synth: new ShellStep("Synthesize", {
        input: CodePipelineSource.connection('owner/group/repo', 'main', {
          connectionArn: 'arn:aws:codestar-connections:us-east-1:222222222222:connection/7d2469ff-514a-4e4f-9003-5ca4a43cdc41',
        }),
        commands: [
          'npm ci',
          'npm run build',
          'npx cdk synth',
        ],
      }),
    });

    pipeline.addStage(new DummyApplication(this, 'DummyApp'));
  }
}
```

### Possible Solution

```diff
--- a/packages/aws-cdk-lib/pipelines/lib/codepipeline/codepipeline-source.ts
+++ b/packages/aws-cdk-lib/pipelines/lib/codepipeline/codepipeline-source.ts
@@ -425,11 +425,11 @@ class CodeStarConnectionSource extends CodePipelineSource 
{
     super(repoString);
 
     const parts = repoString.split('/');
-    if (Token.isUnresolved(repoString) || parts.length !== 2) {
+    if (Token.isUnresolved(repoString) || parts.length < 2) {
       throw new Error(`CodeStar repository name should be a resolved string like '<owner>/<repo>', got '${repoString}'`);
     }
     this.owner = parts[0];
-    this.repo = parts[1];
+    this.repo = parts.slice(1).join('/');
     this.configurePrimaryOutput(new FileSet('Source', this));
   }
```

### Additional Information/Context

_No response_

### CDK CLI Version

2.100.0 (build e1b5c77)

### Framework Version

_No response_

### Node.js Version

v18.17.1

### OS

Ubuntu 20.04.6 LTS

### Language

Python

### Language Version

Python 3.8.10

### Other information

As a workaround until this is fixed, you can manually re-implement [`CodeStarConnectionSource`](https://github.com/aws/aws-cdk/blob/b2a895ef285e5451e64c21e179172e998c479582/packages/aws-cdk-lib/pipelines/lib/codepipeline/codepipeline-source.ts#L420-L450) to include the above changes and instantiate directly.

### PR DESCRIPTION
**Pull request by anshikam**
Closes #27504.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by scanlonp** [comment]
@anshikam, good addition. This will require an integ test. Other than that it looks good!

**Comment by anshikam** [comment]
Thanks @scanlonp for reviewing this :) I can try adding that based on guidance provided in https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md#integration-tests.

**Comment by scanlonp** [comment]
@anshikam you can look here too https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md. I also recently added an example using a pipelines test, so that is (hopefully) doubly helpful!

File: packages/aws-cdk-lib/pipelines/lib/codepipeline/codepipeline-source.ts [hunk 1]
```diff
@@ -256,11 +256,11 @@ class GitHubSource extends CodePipelineSource {
     super(repoString);
 
     const parts = repoString.split('/');
-    if (Token.isUnresolved(repoString) || parts.length !== 2) {
+    if (Token.isUnresolved(repoString) || parts.length < 2) {
       throw new Error(`GitHub repository name should be a resolved string like '<owner>/<repo>', got '${repoString}'`);
```
Comment (scanlonp) [inline packages/aws-cdk-lib/pipelines/lib/codepipeline/codepipeline-source.ts:260]: Lets move this validation into a function. It should cover the edge case of a string such as 'owner/group1//repo' and give a specific error message.

File: packages/aws-cdk-lib/pipelines/lib/codepipeline/codepipeline-source.ts [hunk 2]
```diff
@@ -425,11 +425,11 @@ class CodeStarConnectionSource extends CodePipelineSource {
     super(`${repoString}-${branch}`);
 
     const parts = repoString.split('/');
-    if (Token.isUnresolved(repoString) || parts.length !== 2) {
+    if (Token.isUnresolved(repoString) || parts.length < 2) {
       throw new Error(`CodeStar repository name should be a resolved string like '<owner>/<repo>', got '${repoString}'`);
```
Comment (scanlonp) [inline packages/aws-cdk-lib/pipelines/lib/codepipeline/codepipeline-source.ts:429]: Same thing here.

**Review by scanlonp — CHANGES_REQUESTED** [review]
Other than the integ test, I think we actually need a bit more.

We will need a change to the readme with an example of using a nested repository.
We will need a separate validation function with more descriptive error messages and that catches edge cases.
We will need a test of the validation function with allowed and non-allowed strings.

Each of these should be fairly minor. Thanks!

File: packages/aws-cdk-lib/pipelines/lib/codepipeline/codepipeline-source.ts (hunk 2 above)
Comment (davemeech) [reply packages/aws-cdk-lib/pipelines/lib/codepipeline/codepipeline-source.ts:429]: We might also consider changing the verbiage in the error directory regarding the expected resolved string format.  As an example, the same repository name in the console expresses the format as group/subgroup/project.

**Comment by jgilewski-siili** [comment]
Hi! What is the status of this PR?

**Comment by scanlonp** [comment]
@jgilewski-siili looks like the review comments need to be addressed. If you would like to pick it up, go for it!

**Comment by anshikam** [comment]
I got sidetracked with other work. @jgilewski-siili feel free to take up if you have bandwidth

**Comment by jgilewski-siili** [comment]
@anshikam Yap, the bandwidth is the problem. I will probably have some time in the next week.

**Comment by tim-finnigan** [comment]
@jgilewski-siili is this something you have looked into revisiting?
